### PR TITLE
[-] FO : Fix performances issue with SpecificPrice (PSCSX-6942)

### PR DESCRIPTION
**Pull request by jocel1**
Wait for user feedback before merging :)

### CODE REVIEW

**Comment by drzraf** [comment]
wow! that's not trivial, I'd really like to give you some useful input if I can, but I'll just need a couple of days until I can grab some free time for this.
Right now, and only on the form: why isn't it possible to use the DbQuery SQL builder.
```if(blah) $q->where(foo),```
Since the top level WHERE clause is a sequence of AND-concatenated conditions, this seems possible and would ease readability.

About ``formatIntInQuery()``, tell me if (and how) I'm wrong, but if we only need to search a value of 0, it means that the object ID is NULL (PHP-side), but in this context, does it mean that the database value must be 0 (they can't be NULL)? Couldn't the condition just be stripped-out?

I'll later reply, in the forge, about from/to presence in my own database when I can check this.

I see you're changing the INDEXes, please let me just imagine the following:
- add a boolean "has_specific_price" column to the "product" table
- this boolean is updated when the product is modified and set to TRUE if specific prices affect this product
- when a need for the specific price of a product arise, just check if that boolean is TRUE before doing the search

I understand that it somewhat denormalizes the database by "duplicating" an information, but since this is a costly information (expansive to compute, relying on very dynamic and multiple factors, ...) I think it's worth consideration. For a shop having few specific prices, the gain would be high, for those where specific prices exist for the vast majority of the product, the gain would just be null (but not negative).

Anyway, maybe most specific prices, in the real life, are set according to "group" or a "customer" and not "product", and in such case, adding the boolean to the "group" table or the "customer" table would be better.
Anyway I *guess* that most calls to the getSpecificPrice() are done in a "product-centric" context, and that the more beneficial addition of the above proposed boolean column would be in the "product" table. De facto I see 3 calls to that function in PrestaShop: once in Cart.php (``getProducts``), and twice in Product.php (``priceCalculation()`` and ``isDiscounted()``)

**Comment by jocel1** [comment]
Actually the condition cannot just be stripped-out (I also made this mistake first) : we should exclude what is non 0 (otherwise you could get specific price from non matching products for example), and that's the purpose of the remaining conditions.
Hence your has_specific_price trick doesn't work as well, since even if the product ID is not present in the DB, we should still check global condition (id_product = 0). 
I feel the current approach is a good compromise : if the id_product value is not in the ps_specific_price DB, just do the check with id_product=0. Most of the time, the final condition will be : id_product = 0 AND id_cart = 0 AND id_product_attribute = 0 AND from='0000-00-00 00:00:00' AND to='0000-00-00 00:00:00' which will be efficiently covered by the reordered index on id_product_2, and will be easily cacheable and will not fill the prestashop cache / mysql query cache.
I think I will also add the id_customer in the logic, since it could vary a lot as well.
 
As for the SQL builder, I agree, but it would be too much changes for 1.6.1.x :)

**Comment by jocel1** [comment]
I've just include id_customer in the test, and improve the caching if no specific value is found in the specific_price table

**Comment by drzraf** [comment]
Indeed, I just reviewed the patch and I understand the principles.
I do understand too that having a ``has_specific_price`` on ``{products}`` would not be enough to omit queries. The same flags would have to be added to any table whose entity could be the subject of a specific price (user, group, country, ...). It may not be that practicable (or, maybe, as a cache/temporary table...)

Since the patch relies upon a database upgrade I fear I won't be able to stress it under realistic condition of the production environment [that would be the update more difficult since AFAICT mysql INDEX additions are not conditionally triggered by the upgrade script].

I'll start trying in a dev' environment, and maybe applying the patch without the INDEX to begin with.
But I'm pretty sure that these new INDEX have a potential to speed up things dramatically.

Side note: why ``INDEX (`id_category`, `position`)``  popup in the current PR? It's unrelated, unless I missed something.

File: install-dev/upgrade/sql/1.6.1.2.sql [hunk 1]
```diff
@@ -24,3 +24,11 @@ REPLACE INTO `PREFIX_tag_count` (id_group, id_tag, id_lang, id_shop, counter)
...
 
 ALTER TABLE `PREFIX_search_index` ADD INDEX(`id_product`);
+ALTER TABLE `PREFIX_specific_price` ADD INDEX(`id_product_attribute`);
+ALTER TABLE `PREFIX_specific_price` ADD INDEX(`id_shop`);
+ALTER TABLE `PREFIX_specific_price` ADD INDEX(`id_customer`);
+ALTER TABLE `PREFIX_specific_price` ADD INDEX(`from`);
+ALTER TABLE `PREFIX_specific_price` ADD INDEX(`to`);
+ALTER TABLE `PREFIX_specific_price` DROP KEY `id_product_2`;
+ALTER TABLE `PREFIX_specific_price` ADD UNIQUE KEY `id_product_2` (`id_product`,`id_product_attribute`,`id_customer`,`id_cart`,`from`,`to`,`id_shop`,`id_shop_group`,`id_currency`,`id_country`,`id_group`,`from_quantity`,`id_specific_price_rule`);
+ALTER TABLE `PREFIX_category_product` ADD INDEX(`id_category`, `position`);
```
Comment (inetbiz) [inline install-dev/upgrade/sql/1.6.1.2.sql]: Wrong EOF

**Pull request merged**